python/lsst/summit/utils/tmaUtils.py

```diff
@@ -0,0 +1,1588 @@
+                    print(f"\nThen, applying the {nRowsToApply} rows of data for this event, the state"
+                          " evolved as follows:\n")
+                    firstAppliedRow = False
+
+                # break the row down and print its details
+                rowFor = row['rowFor']
+                axis, rowType = getAxisAndType(rowFor)  # e.g. elevation, MotionState
+                value = tma._getRowPayload(row, rowType, rowFor)
+                valueStr = f"{str(value) if isinstance(value, bool) else value.name}"
+                rowTime = efdTimestampToAstropy(row['private_sndStamp'])
+                print(f"On row {rowNum} the {axis} axis had the {rowType} set to {valueStr} at"
+                      f" {rowTime.utc.isot if useUtc else rowTime.isot}")
+
+                # then apply it as usual, printing the state right afterwards
+                tma.apply(row)
+                tmaStates[rowNum] = tma.state
+                self.printTmaDetailedState(tma)
+                print()
+
+            else:
+                # if it's not in the range of interest then just apply it
+                # silently as usual
+                tma.apply(row)
+                tmaStates[rowNum] = tma.state
+
+    def findEvent(self, time):
+        """Find the event which contains the specified time.
+
+        If the specified time lies within an event, that event is returned. If
+        it is at the exact start, that is logged, and if that start point is
+        shared by the end of the previous event, that is logged too. If the
+        event lies between events, the events either side are logged, but
+        ``None`` is returned. If the time lies before the first event of the
+        day a warning is logged, as for times after the last event of the day.
+
+        Parameters
+        ----------
+        time : `astropy.time.Time`
+            The time.
+
+        Returns
+        -------
+        event : `lsst.summit.utils.tmaUtils.TMAEvent` or `None`
+            The event which contains the specified time, or ``None`` if the
+            time doesn't fall during an event.
+        """
+        # there are four possible cases:
+        # 1) the time lies before the first event of the day
+        # 2) the time lies after the last event of the day
+        # 3) the time lies within an event
+        #   3a) the time is exactly at the start of an event
+        #   3b) if so, time can be shared by the end of the previous event if
+        #       they are contiguous
+        # 4) the time lies between two events
+
+        dayObs = getDayObsForTime(time)
+        # we know this is on the right day, and definitely before the specified
+        # time, but sanity check this before continuing as this needs to be
+        # true for this to give the correct answer
+        assert getDayObsStartTime(dayObs) <= time
+        assert getDayObsEndTime(dayObs) > time
```

Comment:
Perhaps raise a RuntimeError inside an if statement instead of bare asserts.

Reply:
This gets a little philosophical, but in my head at least, `assert` statements are used for things which should logically/definitionally never be possible, and are used to indicate that in code (_i.e._ sanity/consistency checks which should *never* fail, ever, and if they do, some function isn't doing what you thought it did). `if`s and `raise`s in places like that, - again, at least in my mind - are for things which one needs to check, defend against, and raise on when found, so this was why I chose naked `assert`s here rather than check & raise. (Also, I think naked `assert`s, for this reason, can be dropped at runtime, whereas if/raises cannot).